Feature: Add flag for install location (optional)

ArronaxKP opened the pull request:
This change adds a new flag `-i` or `--install` that will allow you to set where the Terraform binary is downloaded/installed into. Useful for edge cases where you need to share commonly installed versions of Terraform. 

This is in draft as it is built upon this PRs that should be merged first. To save headaches with merge conflicts this PR already includes that commit.

- #295 - Thanks @jukie for the PR. 

Once the above PR has been merged ping me and I can update this branch with latest master.

I have created a similar PR for tgswitch here - https://github.com/warrensbox/tgswitch/pull/140

strowi commented:
This would be very useful for CI/CD cases where $HOME/ is not a cache-able path.

ArronaxKP commented:
For CI/CD is what I wrote this for. It turns out we found a workaround for the issue we were having so I didn't need this int eh end

I created a test binary and published it here if you want it but I am not using it myself, I fell back to the original version by warrensbox. I think its just linux_amd64 binary only though:
https://github.com/ArronaxKP/terraform-switcher/releases/tag/tfswitch-with-install-flag

warrensbox commented:
We can revisit this after release v1.10

ArronaxKP commented:
Updated with latest master changes. I have moved away from using tfswitch and tgswitch now. But I wanted to attempt to get this PR updated to as best I can. Since I haven't touched this project in over 9 months, it took a while. I "think" it now works. I tested locally quickly and the `-i` install works however I am using windows so not a real test. I don't have time to fully test it on all platforms. 

If this is still wanted please ping me if you need anything. I will leave this PR in draft as I cannot fully test this locally. Feel free to take the code and create a new branch/PR if that is easier.

yermulnik commented on README.md:
```diff
@@ -7,26 +7,26 @@
-The `tfswitch` command line tool lets you switch between different versions of [Terraform](https://www.terraform.io/).  
-If you do not have a particular version of Terraform installed, `tfswitch` will download and verify the version you desire.  
-The installation is minimal and easy.  
+The `tfswitch` command line tool lets you switch between different versions of [Terraform](https://www.terraform.io/).
+If you do not have a particular version of Terraform installed, `tfswitch` will download and verify the version you desire.
+The installation is minimal and easy.
```
Please put two trailing periods back — they are line breaks in markdown.

ArronaxKP replied:
Ah sorry, I have VS configured to auto remove trailing whitespace. This always catches me out as I avoid the 2 space trailing line breaks due to this break. I tend to use `<br>` instead if I need it based on [line breaks best practices](https://github.com/mattcone/markdown-guide/blob/master/_basic-syntax/line-breaks.md#line-break-best-practices)

Sorry, will put them back now.

ArronaxKP replied:
Fixed

yermulnik commented on README.md:
```diff
@@ -7,26 +7,26 @@
-Going forward we will change the version identifier of `tfswitch` to align with the common go package versioning.  
-Please be advised to change any automated implementation you might have that is relying on the `tfswitch` version string.  
-**Old version string:** `0.1.2412`  
+Going forward we will change the version identifier of `tfswitch` to align with the common go package versioning.
+Please be advised to change any automated implementation you might have that is relying on the `tfswitch` version string.
+**Old version string:** `0.1.2412`
```
Please put two trailing periods back — they are line breaks in markdown.

ArronaxKP replied:
Fixed

yermulnik reviewed (commented):
Thank you for re-visiting this PR.
I do hope you won't be much distracted by re-visiting it once again once the #356 is merged — the #356 brings a lot of refactoring, including the bits touched in this PR.
Thanks in advance.

ArronaxKP commented on lib/install.go:
```diff
@@ -14,7 +14,7 @@ import (
 	installFile               = "terraform"
 	versionPrefix             = "terraform_"
-	installPath               = ".terraform.versions"
+	installFolder             = ".terraform.versions"
```
Actually agree, dir would be better. Will fix this

ArronaxKP commented on www/docs/usage/commandline.md:
```diff
@@ -58,6 +58,23 @@ tfswitch #will automatically switch to terraform version 0.14.4
+**NOTE**
+
+* The folder must exists before you run `tfswitch`
+* The folder passed in `-i`/`--install` must be created before running `tfswtich`
```
Fixed

ArronaxKP commented on README.md:
```diff
@@ -89,4 +89,4 @@ See step-by-step instructions on how to contribute here: [Contribute](https://tf
 ## Issues
-Please open  *issues* here: [New Issue](https://github.com/warrensbox/terraform-switcher/issues)
\ No newline at end of file
+Please open  *issues* here: [New Issue](https://github.com/warrensbox/terraform-switcher/issues)
```
Fixed

ArronaxKP commented:
Resolved all comments. 

If you ping me when https://github.com/warrensbox/terraform-switcher/pull/356 is merged. I will update this again, and re-test it.

yermulnik commented:
> Resolved all comments.

Thank you.

> If you ping me when #356 is merged. I will update this again, and re-test it.

We'll try out best 🤝

yermulnik commented:
> > If you ping me when #356 is merged. I will update this again, and re-test it.
> 
> We'll try out best 🤝

@ArronaxKP Pinging you

ArronaxKP commented:
As promised the PR updated with latest master. Includes the origina `-i/--install` flag changes.

Spotted some other warnings in other files I thought I would fix. 

- Moved homedir function to the `files.go` file inline with all other file functions.
- Fixed markdown spacing issues where there was a missing new line between headers and bullet points.
- Corrected some warnings for booleans based on best practice:
  - [S1008 - Simplify returning boolean expression](https://staticcheck.io/docs/checks#S1008).
  - [S1002 - Omit comparison with boolean constant](https://staticcheck.io/docs/checks#S1002).

My github actions are successful:
![image](https://github.com/warrensbox/terraform-switcher/assets/1584972/a12180b0-fd85-49a2-b113-e20ef87a99d6)

yermulnik commented on lib/param_parsing/parameters.go:
```diff
@@ -34,6 +35,7 @@ func GetParameters() Params {
 	getopt.StringVarLong(&params.CustomBinaryPath, "bin", 'b', "Custom binary path. Ex: tfswitch -b "+lib.ConvertExecutableExt("/Users/username/bin/terraform"))
 	getopt.StringVarLong(&params.DefaultVersion, "default", 'd', "Default to this version in case no other versions could be detected. Ex: tfswitch --default 1.2.4")
 	getopt.BoolVarLong(&params.HelpFlag, "help", 'h', "Displays help message")
+	getopt.StringVarLong(&params.InstallPath, "install", 'i', "Custom install path. Ex: tfswitch -b "+lib.ConvertExecutableExt("/Users/username"))
```
Option description looks a bit messed up with `-b`, doesn't it? 🤔
Also should it probably mention full install path as example for clarity (as in with `installDir` appended)?

ArronaxKP replied:
Will fix/add this now

yermulnik commented on www/docs/usage/commandline.md:
```diff
@@ -17,47 +17,74 @@ The most recently selected versions are presented at the top of the dropdown.
-You can also set the `TF_VERSION` environment variable to your desired terraform version. 
-For example:   
+
+You can also set the `TF_VERSION` environment variable to your desired terraform version.
+For example:
```
I'd glue this together (or put two trailing periods otherwise):
```suggestion
You can also set the `TF_VERSION` environment variable to your desired terraform version. For example:
```

ArronaxKP replied:
Fixing

yermulnik commented on www/docs/usage/commandline.md:
```diff
@@ -17,47 +17,74 @@ The most recently selected versions are presented at the top of the dropdown.
+
+## Install to non-default location
+
+By default `tfswitch` will download the Terraform binary to the user home directory under this path: `/Users/warrenveerasingam/.terraform.versions`
```
It's sort of misleading to use Warren's username and if user is on Linux or Windows. What do you think of this commit suggestion?
```suggestion
By default `tfswitch` will download the Terraform binary to the user home directory under this path: `$HOME/.terraform.versions`
```

ArronaxKP replied:
Fixing

yermulnik commented on www/docs/usage/config-files.md:
```diff
@@ -36,13 +36,13 @@ This is similiar to using a .tfswitchrc file, but you can specify a custom binar
-1. Create a custom binary path. Ex: `mkdir /Users/warrenveerasingam/bin` (replace warrenveerasingam with your username)
-2. Add the path to your PATH. Ex: `export PATH=$PATH:/Users/warrenveerasingam/bin` (add this to your bash profile or zsh profile)
-3. Pass -b or --bin parameter with your custom path to install terraform. Ex: `tfswitch -b /Users/warrenveerasingam/bin/terraform 0.10.8 `
+1. Create a custom binary path. Ex: `mkdir $HOME/bin` (replace warrenveerasingam with your username)
```
```suggestion
1. Create a custom binary path. Ex: `mkdir $HOME/bin`
```

ArronaxKP replied:
Sorry missed that bit fixing.

yermulnik reviewed (commented):
LGTM with minor note below

yermulnik commented:
Also please fetch `master`
![Uploading image.png…]()

yermulnik commented:
@warrensbox @MatrixCrawler @crablab @MatthewJohn @jukie Please review.

ArronaxKP commented:
Updated  with latest master. My Github actions is passing as well

MatrixCrawler reviewed (approved):
LGTM

crablab commented:
Will look this evening/tomorrow morning.

yermulnik commented:
> Will look this evening/tomorrow morning.

I'm giving this my "LGTM". Please merge once you get this reviewed and if you approve when you have a chance later. Thanks.

crablab commented on www/docs/usage/commandline.md:
````diff
@@ -17,47 +17,73 @@ The most recently selected versions are presented at the top of the dropdown.
+tfswitch -i /opt/terraform
+```
+
+**NOTE** - The directory passed in `-i`/`--install` must be created before running `tfswtich`
````
```suggestion
**NOTE** - The directory passed in `-i`/`--install` must be created before running `tfswitch`
```
